File: game_logic/blokus.py
"""Blokus game engine — 20x20 board, 2-4 players, 21 polyomino pieces each."""

PIECES = {
    "I1": [(0, 0)],
    "I2": [(0, 0), (0, 1)],
    "I3": [(0, 0), (0, 1), (0, 2)],
    "V3": [(0, 0), (0, 1), (1, 0)],
    "I4": [(0, 0), (0, 1), (0, 2), (0, 3)],
    "O4": [(0, 0), (0, 1), (1, 0), (1, 1)],
    "T4": [(0, 0), (0, 1), (0, 2), (1, 1)],
    "S4": [(0, 0), (0, 1), (1, 1), (1, 2)],
    "L4": [(0, 0), (0, 1), (0, 2), (1, 0)],
    "F5": [(0, 1), (0, 2), (1, 0), (1, 1), (2, 1)],
    "I5": [(0, 0), (0, 1), (0, 2), (0, 3), (0, 4)],
    "L5": [(0, 0), (0, 1), (0, 2), (0, 3), (1, 0)],
    "N5": [(0, 0), (0, 1), (1, 1), (1, 2), (1, 3)],
    "P5": [(0, 0), (0, 1), (1, 0), (1, 1), (2, 0)],
    "T5": [(0, 0), (0, 1), (0, 2), (1, 1), (2, 1)],
    "U5": [(0, 0), (0, 2), (1, 0), (1, 1), (1, 2)],
    "V5": [(0, 0), (1, 0), (2, 0), (2, 1), (2, 2)],
    "W5": [(0, 0), (1, 0), (1, 1), (2, 1), (2, 2)],
    "X5": [(0, 1), (1, 0), (1, 1), (1, 2), (2, 1)],
    "Y5": [(0, 0), (1, 0), (1, 1), (2, 0), (3, 0)],
    "Z5": [(0, 0), (0, 1), (1, 1), (2, 1), (2, 2)],
}

COLORS = ["#3b82f6", "#eab308", "#ef4444", "#22c55e"]
COLOR_NAMES = ["藍", "黃", "紅", "綠"]
CORNERS = [(0, 0), (0, 19), (19, 19), (19, 0)]
BOARD_SIZE = 20


def normalize(cells):
    min_r = min(r for r, c in cells)
    min_c = min(c for r, c in cells)
    return tuple(sorted((r - min_r, c - min_c) for r, c in cells))


def rotate_90(cells):
    return normalize([(c, -r) for r, c in cells])


def flip_h(cells):
    return normalize([(r, -c) for r, c in cells])


def get_all_orientations(cells):
    orientations = set()
    current = normalize(cells)
    for _ in range(4):
        orientations.add(current)
        orientations.add(flip_h(list(current)))
        current = rotate_90(list(current))
    return [list(o) for o in orientations]


# Pre-compute all orientations for each piece
PIECE_ORIENTATIONS = {}
for name, cells in PIECES.items():
    PIECE_ORIENTATIONS[name] = get_all_orientations(cells)


class BlokusGame:
    def __init__(self, players: list[str]):
        self.players = players
        self.num_players = len(players)
        self.board = [[0] * BOARD_SIZE for _ in range(BOARD_SIZE)]  # 0=empty, 1-4=player
        self.remaining = {i: list(PIECES.keys()) for i in range(self.num_players)}
        self.current_turn = 0
        self.first_move = [True] * self.num_players
        self.passed = [False] * self.num_players
        self.game_over = False
        self.scores = [0] * self.num_players

    def get_state(self, username: str):
        player_idx = self.players.index(username) if username in self.players else -1
        return {
            "board": self.board,
            "players": [
                {
                    "name": self.players[i],
                    "color": COLORS[i],
                    "color_name": COLOR_NAMES[i],
                    "remaining": self.remaining[i],
                    "passed": self.passed[i],
                    "score": self.scores[i],
                }
                for i in range(self.num_players)
            ],
            "current_turn": self.current_turn,
            "my_index": player_idx,
            "game_over": self.game_over,
            "pieces": {name: list(PIECES[name]) for name in PIECES},
            "piece_orientations": {
                name: [list(o) for o in PIECE_ORIENTATIONS[name]]
                for name in PIECES
            },
        }

    def handle_action(self, username: str, action: dict) -> list[dict]:
        player_idx = self.players.index(username)
        if self.game_over:
            return [{"type": "error", "message": "遊戲已結束", "_target": username}]
        if player_idx != self.current_turn:
            return [{"type": "error", "message": "還沒輪到你", "_target": username}]

        action_type = action.get("type")
        if action_type == "place_piece":
            return self._place_piece(player_idx, action)
        elif action_type == "pass":
            return self._pass_turn(player_idx)
        return []

    def _place_piece(self, player_idx: int, action: dict) -> list[dict]:
        piece_name = action.get("piece")
        cells = [tuple(c) for c in action.get("cells", [])]
        username = self.players[player_idx]

        if piece_name not in self.remaining[player_idx]:
            return [{"type": "error", "message": "你沒有這個棋子", "_target": username}]

        # Validate cells match a valid orientation
        normalized = normalize(cells)
        if list(normalized) not in [list(normalize(o)) for o in PIECE_ORIENTATIONS[piece_name]]:
            return [{"type": "error", "message": "棋子形狀不正確", "_target": username}]

        # Validate placement
        error = self._validate_placement(player_idx, cells)
        if error:
            return [{"type": "error", "message": error, "_target": username}]

        # Place the piece
        color = player_idx + 1
        for r, c in cells:
            self.board[r][c] = color
        self.remaining[player_idx].remove(piece_name)
        self.first_move[player_idx] = False

        # Calculate scores
        self._calc_scores()

        events = [{"type": "game_event", "data": {
            "event": "piece_placed",
            "player": username,
            "player_idx": player_idx,
            "piece": piece_name,
            "cells": [list(c) for c in cells],
        }}]

        self._next_turn()

        if self.game_over:
            events.append({"type": "game_event", "data": {
                "event": "game_over",
                "scores": [
                    {"name": self.players[i], "score": self.scores[i]}
                    for i in range(self.num_players)
                ],
            }})

        return events

    def _validate_placement(self, player_idx: int, cells: list[tuple]) -> str | None:
        color = player_idx + 1

        for r, c in cells:
            if r < 0 or r >= BOARD_SIZE or c < 0 or c >= BOARD_SIZE:
                return "超出棋盤範圍"
            if self.board[r][c] != 0:
                return "該位置已有棋子"

        # Check no orthogonal adjacency with same color
        for r, c in cells:
            for dr, dc in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                nr, nc = r + dr, c + dc
                if 0 <= nr < BOARD_SIZE and 0 <= nc < BOARD_SIZE:
                    if self.board[nr][nc] == color and (nr, nc) not in cells:
                        return "不能與自己的棋子邊相鄰"

        if self.first_move[player_idx]:
            corner = CORNERS[player_idx]
            if corner not in cells:
                return f"第一步必須放在角落 ({corner[0]},{corner[1]})"
            return None

        # Check at least one diagonal adjacency with same color
        has_diagonal = False
        for r, c in cells:
            for dr, dc in [(-1, -1), (-1, 1), (1, -1), (1, 1)]:
                nr, nc = r + dr, c + dc
                if 0 <= nr < BOARD_SIZE and 0 <= nc < BOARD_SIZE:
                    if self.board[nr][nc] == color and (nr, nc) not in cells:
                        has_diagonal = True
                        break
            if has_diagonal:
                break

        if not has_diagonal:
            return "必須與自己的棋子角對角相鄰"
        return None

    def _pass_turn(self, player_idx: int) -> list[dict]:
        self.passed[player_idx] = True
        events = [{"type": "game_event", "data": {
            "event": "player_passed",
            "player": self.players[player_idx],
            "player_idx": player_idx,
        }}]
        self._next_turn()
        if self.game_over:
            events.append({"type": "game_event", "data": {
                "event": "game_over",
                "scores": [
                    {"name": self.players[i], "score": self.scores[i]}
                    for i in range(self.num_players)
                ],
            }})
        return events

    def _next_turn(self):
        if all(self.passed):
            self.game_over = True
            self._calc_scores()
            return
        for _ in range(self.num_players):
            self.current_turn = (self.current_turn + 1) % self.num_players
            if not self.passed[self.current_turn]:
                break

    def _calc_scores(self):
        for i in range(self.num_players):
            remaining_squares = sum(
                len(PIECES[p]) for p in self.remaining[i]
            )
            if remaining_squares == 0:
                self.scores[i] = 15
                if len(self.remaining[i]) == 0:
                    last_piece_size = 1  # all placed, check if last was I1
                    self.scores[i] += 5  # bonus for placing all
            else:
                self.scores[i] = -remaining_squares

File: static/js/games/blokus.js
/* Blokus game client */
const CELL = 28;
const BOARD = 20;
const canvas = document.getElementById('boardCanvas');
const ctx = canvas.getContext('2d');
canvas.width = CELL * BOARD + 1;
canvas.height = CELL * BOARD + 1;

let ws, state = null;
let selectedPiece = null;
let selectedColor = null;
let rotation = 0;    // 0-3 (0°, 90°, 180°, 270° CW)
let flipH = false;
let flipV = false;
let previewCells = [];
let mouseCell = null;
let lockedCells = null;
let lockedAnchor = null;
let isDragging = false;
let dragGhost = null;

const COLORS = ['#3b82f6', '#eab308', '#ef4444', '#22c55e'];
const COLORS_LIGHT = ['#93bbfd', '#fde047', '#fca5a5', '#86efac'];
function colorName(i) { return t('blokus.color.' + i); }
const PREVIEW_CELL = 20;

// --- Transform: rotate + flip on client side ---
function transformPiece(baseCells, rot, fH, fV) {
    let cells = baseCells.map(([r, c]) => [r, c]);
    if (fH) cells = cells.map(([r, c]) => [r, -c]);
    if (fV) cells = cells.map(([r, c]) => [-r, c]);
    for (let i = 0; i < rot; i++) {
        cells = cells.map(([r, c]) => [c, -r]);
    }
    const minR = Math.min(...cells.map(([r]) => r));
    const minC = Math.min(...cells.map(([, c]) => c));
    return cells.map(([r, c]) => [r - minR, c - minC]);
}

function currentCells() {
    if (!selectedPiece || !state) return [];
    return transformPiece(state.pieces[selectedPiece], rotation, flipH, flipV);
}

function pieceOffset(cells) {
    if (!cells.length) return [0, 0];
    const maxR = Math.max(...cells.map(([r]) => r));
    const maxC = Math.max(...cells.map(([, c]) => c));
    return [Math.floor(maxR / 2), Math.floor(maxC / 2)];
}

// --- WebSocket ---
function connect() {
    const proto = location.protocol === 'https:' ? 'wss' : 'ws';
    ws = new WebSocket(`${proto}://${location.host}/ws/${ROOM_ID}`);
    ws.onmessage = e => {
        const msg = JSON.parse(e.data);
        if (msg.type === 'chat') handleChatMessage(msg.data);
        else handleMessage(msg);
    };
    ws.onclose = () => setTimeout(connect, 2000);
    ws.onopen = () => initChat(ws);
}

function handleMessage(msg) {
    if (msg.type === 'game_state') {
        state = msg.data;
        render();
    } else if (msg.type === 'error') {
        showMessage(msg.message, 'lose');
    } else if (msg.type === 'game_event') {
        const d = msg.data;
        if (d.event === 'piece_placed') {
            lockedCells = null; lockedAnchor = null; selectedPiece = null; selectedColor = null;
            document.getElementById('confirmBar').style.display = 'none';
        } else if (d.event === 'game_over') {
            const scores = d.scores.sort((a, b) => b.score - a.score);
            showMessage(t('blokus.champion', {name: scores[0].name, score: scores[0].score}), 'win');
        }
    }
}

// --- Validation (client-side, mirrors server rules) ---
function isValidPlacement(cells) {
    if (!state || selectedColor === null) return false;
    const bc = selectedColor + 1;
    const cellSet = new Set(cells.map(([r,c]) => `${r},${c}`));
    for (const [r, c] of cells) {
        if (r < 0 || r >= BOARD || c < 0 || c >= BOARD) return false;
        if (state.board[r][c] !== 0) return false;
    }
    for (const [r, c] of cells) {
        for (const [dr, dc] of [[-1,0],[1,0],[0,-1],[0,1]]) {
            const nr = r+dr, nc = c+dc;
            if (nr>=0 && nr<BOARD && nc>=0 && nc<BOARD)
                if (state.board[nr][nc] === bc && !cellSet.has(`${nr},${nc}`)) return false;
        }
    }
    if (state.first_move[selectedColor]) {
        return cells.some(([r,c]) =>
            state.available_corners.some(([cr,cc]) => r===cr && c===cc));
    }
    for (const [r, c] of cells) {
        for (const [dr, dc] of [[-1,-1],[-1,1],[1,-1],[1,1]]) {
            const nr = r+dr, nc = c+dc;
            if (nr>=0 && nr<BOARD && nc>=0 && nc<BOARD)
                if (state.board[nr][nc] === bc && !cellSet.has(`${nr},${nc}`)) return true;
        }
    }
    return false;
}

function isMyTurn() { return state && state.is_my_turn && !state.game_over; }
function curColor() { return state ? state.current_color : 0; }

// --- Render ---
function render() {
    if (!state) return;
    drawBoard();
    drawPlayers();
    drawPiecesPanel();
    drawPreviewArea();
    updateTurnInfo();
}

function drawBoard() {
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    for (let r = 0; r < BOARD; r++) {
        for (let c = 0; c < BOARD; c++) {
            const v = state.board[r][c];
            ctx.fillStyle = v > 0 ? COLORS[v - 1] : '#f8fafc';
            ctx.fillRect(c * CELL, r * CELL, CELL, CELL);
            ctx.strokeStyle = '#e2e8f0';
            ctx.strokeRect(c * CELL, r * CELL, CELL, CELL);
        }
    }
    // Available corners highlight
    if (selectedColor !== null && state.first_move[selectedColor]) {
        state.available_corners.forEach(([r, c]) => {
            ctx.fillStyle = COLORS[selectedColor];
            ctx.globalAlpha = 0.2;
            ctx.fillRect(c * CELL, r * CELL, CELL, CELL);
            ctx.globalAlpha = 1;
            ctx.strokeStyle = COLORS[selectedColor];
            ctx.lineWidth = 2;
            ctx.strokeRect(c * CELL + 1, r * CELL + 1, CELL - 2, CELL - 2);
            ctx.lineWidth = 1;
        });
    }

    // Locked placement
    if (lockedCells) {
        drawCellsOnBoard(lockedCells, selectedColor, 0.65, true);
        return;
    }
    // Hover/drag preview
    if (previewCells.length > 0 && isValidPlacement(previewCells)) {
        drawCellsOnBoard(previewCells, selectedColor, 0.4, false);
    }
}

function drawCellsOnBoard(cells, colorIdx, alpha, border) {
    cells.forEach(([r, c]) => {
        if (r < 0 || r >= BOARD || c < 0 || c >= BOARD) return;
        ctx.fillStyle = COLORS[colorIdx] || COLORS[0];
        ctx.globalAlpha = alpha;
        ctx.fillRect(c * CELL, r * CELL, CELL, CELL);
        ctx.globalAlpha = 1;
        if (border) {
            ctx.strokeStyle = '#fff';
            ctx.lineWidth = 2;
            ctx.strokeRect(c * CELL + 1, r * CELL + 1, CELL - 2, CELL - 2);
            ctx.lineWidth = 1;
        }
    });
}

function drawPlayers() {
    document.getElementById('playersPanel').innerHTML = state.players.map((p, i) => {
        const dots = p.colors.map(c =>
            `<span class="color-dot${c===state.current_color?' active-dot':''}" style="background:${COLORS[c]}"></span>`
        ).join('');
        return `<div class="player-info ${i===state.current_player_idx?'active':''}">
            <div class="player-header">${dots} <strong>${p.name}${i===state.my_index?` (${t('lobby.you')||'你'})`:''}</strong></div>
            <div>${p.score} ${t('blokus.score')}</div>
        </div>`;
    }).join('');
    // Mobile bar
    const mb = document.getElementById('blokusMobilePlayers');
    if (mb) mb.innerHTML = state.players.map((p, i) => {
        const dots = p.colors.map(c => `<span style="color:${COLORS[c]}">●</span>`).join('');
        return `<span class="mp-chip ${i===state.current_player_idx?'active':''}">${dots} ${p.name} ${p.score}</span>`;
    }).join('');
}

// --- Pieces Panel: always show player's colors ---
function drawPiecesPanel() {
    const panel = document.getElementById('piecesPanel');
    if (!state || state.my_index < 0) { panel.innerHTML = ''; return; }

    const s = 9;
    let html = '';
    for (const ci of state.my_colors) {
        const info = state.color_info[ci];
        const isCurrent = (ci === state.current_color);
        html += `<div class="color-section ${isCurrent ? 'current-color' : ''}">
            <div class="color-section-hdr">
                <span class="color-dot" style="background:${COLORS[ci]}"></span>
                <span>${colorName(ci)}</span>
                <span class="piece-count">${info.remaining.length}</span>
                ${isCurrent ? '<span class="now-badge">NOW</span>' : ''}
            </div>
            <div class="color-pieces">`;
        for (const name of info.remaining) {
            const cells = state.pieces[name];
            const maxR = Math.max(...cells.map(c => c[0])) + 1;
            const maxC = Math.max(...cells.map(c => c[1])) + 1;
            const isSel = selectedPiece === name && selectedColor === ci;
            html += `<div class="piece-item ${isSel ? 'selected' : ''}" onclick="selectPiece('${name}',${ci})">
                <svg width="${maxC*s+2}" height="${maxR*s+2}" viewBox="0 0 ${maxC*s+2} ${maxR*s+2}">
                    ${cells.map(([r, c]) =>
                        `<rect x="${c*s+1}" y="${r*s+1}" width="${s}" height="${s}" fill="${COLORS[ci]}" stroke="#fff" stroke-width="0.5"/>`
                    ).join('')}
                </svg>
            </div>`;
        }
        html += '</div></div>';
    }
    panel.innerHTML = html;
}

// --- Preview Area ---
function drawPreviewArea() {
    const area = document.getElementById('previewArea');
    if (!selectedPiece || !state) {
        area.innerHTML = `<div class="preview-hint">${t('blokus.select')}</div>`;
        return;
    }
    const cells = currentCells();
    const svg = cellsToSVG(cells, PREVIEW_CELL, selectedColor);
    area.innerHTML = `<div class="preview-piece" id="previewPiece">${svg}</div>
        <div class="preview-drag-hint">${t('blokus.drag_hint')}</div>`;
    const el = document.getElementById('previewPiece');
    el.addEventListener('mousedown', e => { e.preventDefault(); startDrag(e.clientX, e.clientY); });
    el.addEventListener('touchstart', e => { e.preventDefault(); startDrag(e.touches[0].clientX, e.touches[0].clientY); }, { passive: false });
}

function cellsToSVG(cells, s, colorIdx) {
    const maxR = Math.max(...cells.map(([r]) => r)) + 1;
    const maxC = Math.max(...cells.map(([, c]) => c)) + 1;
    return `<svg width="${maxC*s+2}" height="${maxR*s+2}" viewBox="0 0 ${maxC*s+2} ${maxR*s+2}" class="preview-svg">
        ${cells.map(([r, c]) =>
            `<rect x="${c*s+1}" y="${r*s+1}" width="${s}" height="${s}" rx="2"
                   fill="${COLORS[colorIdx]}" stroke="#fff" stroke-width="1.5"/>`
        ).join('')}
    </svg>`;
}

// --- Selection & Orientation ---
function selectPiece(name, colorIdx) {
    if (lockedCells) return;
    selectedPiece = name;
    selectedColor = colorIdx;
    rotation = 0; flipH = false; flipV = false;
    previewCells = [];
    drawPiecesPanel();
    drawPreviewArea();
    drawBoard();
}

function rotatePiece() {
    if (!selectedPiece) return;
    rotation = (rotation + 1) % 4;
    afterOrientationChange();
}

function flipHorizontal() {
    if (!selectedPiece) return;
    flipH = !flipH;
    afterOrientationChange();
}

function flipVertical() {
    if (!selectedPiece) return;
    flipV = !flipV;
    afterOrientationChange();
}

function afterOrientationChange() {
    if (lockedCells && lockedAnchor) {
        const cells = currentCells();
        lockedCells = cells.map(([r, c]) => [r + lockedAnchor[0], c + lockedAnchor[1]]);
        if (!isValidPlacement(lockedCells)) {
            lockedCells = null; lockedAnchor = null;
            document.getElementById('confirmBar').style.display = 'none';
        }
    }
    drawPreviewArea();
    updateHoverPreview();
    drawBoard();
}

function updateHoverPreview() {
    if (!selectedPiece || !mouseCell || !state || lockedCells) {
        if (!lockedCells) previewCells = [];
        return;
    }
    const cells = currentCells();
    const [offR, offC] = pieceOffset(cells);
    previewCells = cells.map(([r, c]) => [r + mouseCell[0] - offR, c + mouseCell[1] - offC]);
}

// --- Board coord helper ---
function canvasCellFromPoint(x, y) {
    const rect = canvas.getBoundingClientRect();
    if (x < rect.left || x > rect.right || y < rect.top || y > rect.bottom) return null;
    const sx = canvas.width / rect.width;
    const sy = canvas.height / rect.height;
    const c = Math.floor((x - rect.left) * sx / CELL);
    const r = Math.floor((y - rect.top) * sy / CELL);
    if (r >= 0 && r < BOARD && c >= 0 && c < BOARD) return [r, c];
    return null;
}

// --- Drag ---
function startDrag(x, y) {
    if (!selectedPiece || lockedCells) return;
    isDragging = true;
    canvas.classList.add('drop-target');
    dragGhost = document.createElement('div');
    dragGhost.className = 'drag-ghost';
    dragGhost.innerHTML = cellsToSVG(currentCells(), 16, selectedColor);
    document.body.appendChild(dragGhost);
    positionGhost(x, y);
}

function positionGhost(x, y) {
    if (!dragGhost) return;
    const w = dragGhost.offsetWidth || 40;
    const h = dragGhost.offsetHeight || 40;
    dragGhost.style.left = (x - w / 2) + 'px';
    dragGhost.style.top = (y - h / 2) + 'px';
}

function endDrag(x, y) {
    isDragging = false;
    canvas.classList.remove('drop-target');
    if (dragGhost) { dragGhost.remove(); dragGhost = null; }
    const cell = canvasCellFromPoint(x, y);
    if (cell) {
        mouseCell = cell;
        updateHoverPreview();
        if (previewCells.length > 0 && isValidPlacement(previewCells)) lockPlacement();
    }
    previewCells = [];
    drawBoard();
}

document.addEventListener('mousemove', e => {
    if (isDragging) {
        positionGhost(e.clientX, e.clientY);
        const cell = canvasCellFromPoint(e.clientX, e.clientY);
        if (cell) { mouseCell = cell; updateHoverPreview(); drawBoard(); }
        else { previewCells = []; drawBoard(); }
    } else if (!lockedCells && selectedPiece) {
        const cell = canvasCellFromPoint(e.clientX, e.clientY);
        if (cell) { mouseCell = cell; updateHoverPreview(); drawBoard(); }
    }
});
document.addEventListener('mouseup', e => { if (isDragging) endDrag(e.clientX, e.clientY); });
document.addEventListener('touchmove', e => {
    if (!isDragging) return;
    const t = e.touches[0];
    positionGhost(t.clientX, t.clientY);
    const cell = canvasCellFromPoint(t.clientX, t.clientY);
    if (cell) { mouseCell = cell; updateHoverPreview(); drawBoard(); }
    else { previewCells = []; drawBoard(); }
}, { passive: false });
document.addEventListener('touchend', e => {
    if (!isDragging) return;
    endDrag(e.changedTouches[0].clientX, e.changedTouches[0].clientY);
});

canvas.addEventListener('mouseleave', () => {
    if (lockedCells || isDragging) return;
    mouseCell = null; previewCells = []; drawBoard();
});

canvas.addEventListener('click', () => {
    if (lockedCells || isDragging) return;
    if (previewCells.length > 0 && isValidPlacement(previewCells)) lockPlacement();
});

// --- Lock / Confirm / Cancel ---
function lockPlacement() {
    if (!selectedPiece || previewCells.length === 0) return;
    lockedCells = [...previewCells];
    if (mouseCell) {
        const cells = currentCells();
        const [offR, offC] = pieceOffset(cells);
        lockedAnchor = [mouseCell[0] - offR, mouseCell[1] - offC];
    } else {
        lockedAnchor = null;
    }
    document.getElementById('confirmBar').style.display = 'flex';
    drawBoard();
}

function confirmPlace() {
    if (!lockedCells || !selectedPiece || selectedColor === null) return;
    if (selectedColor !== curColor()) {
        showMessage(t('blokus.not_color', {color: colorName(selectedColor)}), 'lose');
        return;
    }
    if (!isMyTurn()) {
        showMessage(t('blokus.not_you'), 'lose');
        return;
    }
    ws.send(JSON.stringify({ type: 'place_piece', piece: selectedPiece, cells: lockedCells }));
}

function cancelPlace() {
    lockedCells = null; lockedAnchor = null;
    document.getElementById('confirmBar').style.display = 'none';
    drawBoard();
}

function passTurn() {
    if (!isMyTurn()) return;
    ws.send(JSON.stringify({ type: 'pass' }));
}

// --- UI ---
function updateTurnInfo() {
    const el = document.getElementById('turnInfo');
    if (!state) return;
    if (state.game_over) { el.textContent = t('blokus.game_over'); return; }
    const ci = curColor();
    const dot = `<span class="color-dot" style="background:${COLORS[ci]}"></span>`;
    if (isMyTurn()) {
        el.innerHTML = `${dot} <strong style="color:${COLORS[ci]}">${t('blokus.your_turn', {color: colorName(ci)})}</strong>`;
    } else {
        el.innerHTML = `${dot} ${t('blokus.wait_turn', {name: state.players[state.current_player_idx].name, color: colorName(ci)})}`;
    }
}

function showMessage(text, cls) {
    const el = document.getElementById('gameMessage');
    el.textContent = text;
    el.className = `game-message ${cls || ''}`;
    if (cls) setTimeout(() => { el.textContent = ''; el.className = 'game-message'; }, 3000);
}

document.addEventListener('keydown', e => {
    if (e.key === 'r' || e.key === 'R') rotatePiece();
    if (e.key === 'f' || e.key === 'F') flipHorizontal();
    if (e.key === 'v' || e.key === 'V') flipVertical();
    if (e.key === 'Escape') cancelPlace();
});

connect();
